usb: rework UHC transfers and add initial tests for new device support

jfischer-no opened the pull request:
Rework UHC transfers and add initial tests for new device support.
Tested on nRF52840DK+MAX3421e and qemu_cortex_m3.

jfischer-no committed:
tests: usb: add test for new USB device support

This is initial patch to add tests for new USB device support.
The test uses USB host support and virtual USB bus by default,
but should work on real hardware as well. For now, only the
Get Configuration and Set Interface requests are tested by default.
More tests will follow.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: add a structure to represent a USB device.

Add a structure to represent a USB device and wrappers to avoid
glue UHC calls when operating on devices. Although there is a long
road to device configuration and management, we can start with
the tools that can also be used for USB device support testing.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usbh: remove usbh_peripheral and rename class_code

Rename usbh_class_code to usbh_code_triple and remove reserved
member to bettter reflect the purpose of this structure.
Remove unused struct usbh_peripheral.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: move transfer status to transfer structure

Aling with the changes in UDC done in the commit ad81b3b79747
("drivers: udc: move transfer status to buffer info structure")
This allows us to get the result of synchronous transfer and
simplify uhc_submit_event().

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: rework transfer buffer handling

The current approach is a bit impractical in the upper layer.
This patch removes the two fifos that hold the transfer buffers
and replaces them with a byte array for the setup packet and
a pointer to a data buffer. The data buffer is mandatory for
all types of transfers except control without a data stage.
The waste of eight unused bytes for non-control transfers should
be insignificant, since an additional pointer would be at least
half of it, and then there would be the overhead of handling it.

This patch also clean up the transfer flags, rename owner to callback
as it reflects the upper layer use case, and add an additional member
to hold the pointer to the USB device (peripheral on the bus).

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: loopback: fixup left over from early state

Class instance must not call usbd_ep_ctrl_enqueue().

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: cleanup UHC event processing

Remove ugly hack to call class driver ops.
This is preparation for later changes where we will change
parts to work on structure that represents USB device and
not on glue UHC functions.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: rework ch9 tools to work on USB device

Use the USB device structure and wrappers introduced earlier.
With this patch, users of ch9 requests do not need to work
directly with the UHC API. The requests are now blocked until
there is a response from the UHC. Callers finally have access
to the data buffer and request status.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: remove unused USBH_DEFINE_CLASS(bazfoo) from the shell

With the latest change, there is no need for this structure
to be used as a completion handler.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: add command to get the current device configuration

Add command to get the current device configuration.
Revise the shell part to have a set|get configuration
subcommands.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/host/usbh_ch9.c:
```diff
@@ -60,12 +54,11 @@ int usbh_req_setup(const struct device *dev,
 		if (usb_reqtype_is_to_device(&req) && data != NULL) {
```
While this commit does not modify this particular condition, I think it might be good to assert (outside this function) on `(data != NULL && wLength) || (data == NULL && !wLength)`. Then we could remove the check against the `data != NULL) from here.

jfischer-no replied:
Not sure where ASSERT should be placed, all the callers are simplified, I can imagine to add
`__ASSERT((buf != NULL && wLength) || (buf == NULL && !wLength), "Blabla");` just after memcpy().

tmon-nordic replied:
Just after memcpy() is fine.

tmon-nordic commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -107,27 +107,14 @@ static int bazfoo_request(struct usbh_contex *const ctx,
-
-		buf = net_buf_get(&xfer->done, K_NO_WAIT);
-		if (buf) {
-			/*
-			 * FIXME: We don not distinguish the context
-			 * of the request and always try to print it
-			 * as descriptor first. If it is not a known descriptor,
-			 * we show a hexdump in any case.
-			 * This is just simple enough for first steps and will
-			 * be revised with coming peripheral device management.
-			 */
```
I think this comment should stay because it still holds.

jfischer-no replied:
Actually it should be removed in the commit "usb: host: rework ch9 tools to work on USB device" not here.

tmon-nordic commented on subsys/usb/host/usbh_ch9.c:
```diff
@@ -33,139 +45,188 @@ int usbh_req_setup(const struct device *dev,
-		}
-
-		ret = uhc_xfer_buf_add(dev, xfer, buf);
+	if (wLength && buf != NULL) {
```
I would prefer to have assert for `(data != NULL && wLength) || (data == NULL && !wLength)` before the if condition and then just check `wLength` inside the if. I prefer invalid code to crash on NULL pointer somewhere (if asserts are disabled) than to have the code silence the error and propagate it further down.

The way this is currently implemented, it is possible for invalid code (i.e. code that calls this function with clearly wrong parameters) to send wLength > 0 *and* have the host to proceed directly to status stage. Please make the future debugging easier and just assert.

jfischer-no replied:
I will add commit to address it.

tmon-nordic commented on subsys/usb/host/usbh_ch9.c:
```diff
@@ -33,139 +45,188 @@ int usbh_req_setup(const struct device *dev,
+		goto buf_alloc_err;
+	}
+
+	k_sem_take(&ch9_req_sync, K_MSEC(100));
```
Why 100 ms? `ch9_req_sync` is given after the control transfer finishes or did I skip something important? If yes, then this is lower than what USB 2.0 specification allows. Also blocking here really kills performance because it prohibits other transfers from executing (for one device this is not such a big deal, but also e.g. prevents bulk transfer to execute alongside control transfer).

jfischer-no replied:
Semaphore should only be used for synchronous transfer, I would like to move it to `struct usb_device{}`, but then additional work is also needed. This is on my list.

tmon-nordic commented on subsys/usb/host/usbh_shell.c:
```diff
@@ -560,54 +563,54 @@ static int cmd_usbh_disable(const struct shell *sh,
+	SHELL_CMD_ARG(bulk, NULL, "<endpoint> <length>",
+		      cmd_bulk, 3, 0),
 	SHELL_SUBCMD_SET_END
 );
```
Why are you removing `<address>` from the commands? Even with current just one device it is useful to be able to address packet to non-existent device to check whether the connected device really does not care about data addressed to others.

jfischer-no replied:
I agree that we need this kind of commands or a way to test this request (must have).  But the address argument here is in the way to have device management. Thoughts: Inside struct usb_device there would be a path for the device, e.g. controller->hub->...->device. Now how to perform "row" request on the specific bus, one way is to have a dummy device with a path, the other is to submit "row" requests on specific controller bus. Anyway, this kind of commands should probably be for testing only and disabled by default. I am looking for a solution but unlikely it can be part or this PR.

tmon-nordic replied:
BUt how can you have device management if you cannot select the device you want to communicate with? In general USB is sending the data to all downstream ports, only the FS/LS devices connected via HS hub do not get other devices data (SPLIT token includes the port number where the FS/LS packets are supposed to go).

Because there can be only up to 5 hubs in between host and device, you can uniquely identify every device by knowing to which host controller it is connected to and the 5 bytes that form the path.

jfischer-no replied:
But selecting a device, e.g. in the shell (in the future, it  would required some parts of device management be disabled) / submit a request to this device (the path to a device would be inside struct usb_device{}), and submit a specific request using a controller/bus are different use cases.

jfischer-no commented on drivers/usb/uhc/uhc_common.c:
```diff
@@ -81,13 +80,38 @@ int uhc_xfer_append(const struct device *dev,
+		return buf;
+	}
+
+	if (buf->size < size) {
+		LOG_ERR("Buffer is smaller than requested");
+		net_buf_unref(buf);
+		buf = NULL;
+	}
```
I will remove it. This is the same pattern as in UDC, IIR I was too paranoid in the past when I tried different pools.

jfischer-no committed:
usb: host: usbh_ch9: add ASSERT for unresolved data stage conditions

Do not explicitly check buf parameter in usbh_req_setup()
but add ASSERT to check unresolved data stage conditions.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic reviewed (dismissed):
I would prefer address to stay in the commands but how to support hubs is still an open question.
